pkg/ccl/changefeedccl/sink.go

```diff
@@ -587,6 +639,256 @@ func (s *bufferSink) Close() error {
 	return nil
 }
 
+// cloudStorageFormatBucket formats times as YYYYMMDDHHMMSSNNNNNNNNN.
+func cloudStorageFormatBucket(t time.Time) string {
+	// TODO(dan): Instead do the minimal thing necessary to differentiate times
+	// truncated to some bucket size.
+	const f = `20060102150405`
+	return fmt.Sprintf(`%s%09d`, t.Format(f), t.Nanosecond())
+}
+
+type cloudStorageSinkKey struct {
+	Bucket   time.Time
+	Topic    string
+	SchemaID sqlbase.DescriptorVersion
+	SinkID   string
+	Ext      string
+}
+
+func (k cloudStorageSinkKey) Filename() string {
+	return fmt.Sprintf(`%s-%s-%d-%s%s`,
+		cloudStorageFormatBucket(k.Bucket), k.Topic, k.SchemaID, k.SinkID, k.Ext)
+}
+
+// cloudStorageSink emits to files on cloud storage.
+//
+// The data files are named `<timestamp>_<topic>_<schema_id>_<uniquer>.<ext>`.
+//
+// `<timestamp>` is truncated to some bucket size, specified by the required
+// sink param `bucket_size`. Bucket size is a tradeoff between number of files
+// and the end-to-end latency of data being resolved.
+//
+// `<topic>` corresponds to one SQL table.
+//
+// `<schema_id>` changes whenever the SQL table schema changes, which allows us
+// to guarantee to users that _all entries in a given file have the same
+// schema_.
+//
+// `<uniquer>` is used to keep nodes in a cluster from overwriting each other's
+// data and should be ignored by external users.
```

Comment:
Does the uniquer also keep a single node from overwriting its own data if there are multiple changefeeds, or if a changefeed gets cancelled/restarted?